api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+	// Deny: denies the selected traffic
+	// Pass: instructs the selected traffic to skip any remaining positive priority
+	// ANP rules, and then pass execution to any NetworkPolicies that select the pod.
+	// If the pod is not selected by any NetworkPolicies then execution
+	// is passed to any Priority=0 ANP rules that select the pod, whereupon the traffic
+	// is allowed if not dropped by any Priority=0 ANP rules.
+	// This field is mandatory.
+	Action AdminNetPolRuleAction `json:"action"`
+
+	// Ports allows for matching on traffic based on port and protocols.
+	// This field is mandatory.
+	Ports AdminNetworkPolicyPorts `json:"ports"`
+
+	// List of destinations to which traffic will be allowed/denied/passed from the entities
+	// selected by this AdminNetworkPolicyRule. Items in this list are combined using a logical OR
+	// operation. If this field is empty, this rule matches no destinations.
+	// If this field is present and contains at least one item, this rule
+	// allows/denies/passes traffic to the defined AdminNetworkPolicyPeer(s)
+	// If it is empty no traffic is matched by the AdminNetworkPolicyEgressRule.
+	To []AdminNetworkPolicyPeer `json:"to"`
+}
+
+// AdminNetworkPolicyPorts handles selection of traffic based on port(s).
+// Exactly one of the fields must be defined.
+type AdminNetworkPolicyPorts struct {
+	// AllPorts cannot be "false" when it is set
+	// AllPorts allows the user to select all ports for all protocols, thus not
+	// selecting traffic based on L4 principles.
+	// If "true" then all ports are selected for all protocols.
+	// +optional
+	AllPorts *bool `json:"allPorts,omitempty"`
+
+	// The list of ports to allow/deny/pass traffic on, each item in this list is
+	// combined using a logical OR. When this field is present it should contain at
+	// least one item, and this rule allows/denies/passes traffic only if the traffic
+	// matches at least one port in the list.
+	// +optional
+	List []AdminNetworkPolicyPort `json:"list,omitempty"`
+}
+
+// AdminNetworkPolicyPort describes a port to select
+type AdminNetworkPolicyPort struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
+	// field defaults to TCP.
+	// +optional
+	// +kubebuilder:default=TCP
+	Protocol *v1.Protocol `json:"protocol,omitempty"`
+
+	// The port on the given protocol. This can either be a numerical or named
+	// port on a pod. If this field is not provided, this matches no port names and
+	// numbers.
+	// If present, only traffic on the specified protocol AND port will be matched.
+	// +optional
+	Port *intstr.IntOrString `json:"port,omitempty"`
+
+	// If set, indicates that the range of ports from port to endPort, inclusive,
+	// should be allowed by the policy. This field cannot be defined if the port field
+	// is not defined or if the port field is defined as a named (string) port.
+	// The endPort must be equal or greater than port.
+	// +optional
+	EndPort *int32 `json:"endPort,omitempty"`
```

Comment:
We did it this way in NP because we had compatibility constraints, but we don't have to copy the same behavior in ANP, especially if it requires weird non-kubebuilder-validatable constraints like "`endPort` can only be specified if you didn't use a named port".

There's also another broken part (both here and in NP) which is that you shouldn't be allowed to specify `protocol` if you specify a named port, since the named port already implies a protocol.

So there should be three possibilities here:
- named port
- numeric port plus optional protocol
- numeric port range plus optional protocol

and we probably don't want the latter as `port: 6000, endPort: 6100`. Any of these would be better:

    portRange: 6000-6100

    portRange:
      start: 6000
      end: 6100

    startPort: 6000
    endPort: 6100

Reply:
I agree this could use some thought, will leave this comment up for further conversation